### PR TITLE
fix(animations): guard against views being removed/inserted quickly

### LINKED ISSUE #19712: bug: Cannot read property insertNode of undefine
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
</code></pre>

## Current behavior

Sorry my english is not good but I try my best to describe the problem.

Navigating by back and forward to component with directive and animation cause error.

## Expected behavior

No error occur.

## Minimal reproduction of the problem with instructions

https://github.com/sandangel/angular-bug.git

yarn
yarn start

click on table link, then then graph link will show up. Click on it and the graph will show up, use back and forward button on mouse to navigate back and forward quickly between table and graph. Keep doing it several times and the bug will occur.

## Environment

<pre><code>
@angular/cli: 1.4.7
node: 8.7.0
os: darwin x64
@angular/animations: 4.4.5
@angular/cdk: 2.0.0-beta.12
@angular/common: 4.4.5
@angular/compiler: 4.4.5
@angular/core: 4.4.5
@angular/flex-layout: 2.0.0-beta.9
@angular/forms: 4.4.5
@angular/http: 4.4.5
@angular/material: 2.0.0-beta.12
@angular/platform-browser: 4.4.5
@angular/platform-browser-dynamic: 4.4.5
@angular/router: 4.4.5
@angular/cli: 1.4.7
@angular/compiler-cli: 4.4.5
@angular/language-service: 4.4.5
typescript: 2.5.3

this happens in angular 5.0.0-rc.2 too

Browser:
- [x] Chrome (desktop) version 61
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.7.0  
- Platform:   MacOS High Sierra, Linux x64

Others:

</code></pre>
Actually I sent a pull request to fix this problem but seem like no one noticed it so I post an issue here. I just add 

```javascript
if (this._fetchNamespace(namespaceId)) {
    this._fetchNamespace(namespaceId).insertNode(element, parent);
}
```
and the bug has gone.

https://github.com/angular/angular/pull/19662

### PR DESCRIPTION
**Pull request by matsko**
Closes #19712

### CODE REVIEW

**Review by mhevery — CHANGES_REQUESTED** [review]
We need a test. :-(

**Comment by sandangel** [comment]
ping @matsko

**Comment by kenneth-igaw** [comment]
+1

**Comment by KennethanCeyer** [comment]
+1

----

I am suffering from the same error.
This error is also occurring in angular release 5.0.

**Comment by sandangel** [comment]
The problem here is how we can write a test for this case

**Comment by sandangel** [comment]
@matsko can you review this?

**Comment by KrzysztofJaneczko** [comment]
+1 - I quite badly need these changes.

**Comment by sandangel** [comment]
seem like no one is looking into this issue now

**Comment by sandangel** [comment]
@KrzysztofJaneczko can you reproduce the issue with my stackblitz link? Seem like they can not reproduce the issue with my reproduce steps. Can you provide another reproduce steps then they can can consider fix it quickly.

File: packages/animations/browser/src/render/transition_animation_engine.ts [hunk 1]
```diff
@@ -659,7 +659,14 @@ export class TransitionAnimationEngine {
     // code does not contain any animation code in it, but it is
     // just being called so that the node is marked as being inserted
     if (namespaceId) {
-      this._fetchNamespace(namespaceId).insertNode(element, parent);
+      // in a rare situation the namespace might've already been
+      // removed before a insertion has occured (due to a conflict)
+      // between the regsiter/deregistration of a view therefore it's
+      // important to check to see if the NS has been returned
+      const ns = this._fetchNamespace(namespaceId);
```
Comment (JoostK) [inline packages/animations/browser/src/render/transition_animation_engine.ts]: Would it be fair to let `_fetchNamespace` include `undefined` in its return type, for the type checker to be able to check this? There's quite some more unchecked usages of `_fetchNamespace` and also various direct lookups into `_namespaceLookup`

File: packages/animations/browser/src/render/transition_animation_engine.ts [hunk 2]
```diff
@@ -659,7 +659,14 @@ export class TransitionAnimationEngine {
     // code does not contain any animation code in it, but it is
     // just being called so that the node is marked as being inserted
     if (namespaceId) {
-      this._fetchNamespace(namespaceId).insertNode(element, parent);
+      // in a rare situation the namespace might've already been
+      // removed before a insertion has occured (due to a conflict)
+      // between the regsiter/deregistration of a view therefore it's
```
Comment (JoostK) [inline packages/animations/browser/src/render/transition_animation_engine.ts]: typo: *regsiter* -> *registration*

**Comment by mikeagarza** [comment]
Randomly running into this issue as well.  Would like to see this fix go in soon.

**Comment by mhevery** [comment]
This can not land until we have a test for this fix.

**Comment by matsko** [comment]
Closing this in favor of https://github.com/angular/angular/pull/21977

**Pull request closed without merging**